docs/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to.md

```diff
@@ -0,0 +1,500 @@
+
+* Attribute on property
+* <xref:System.Text.Json.JsonSerializerOptions.Converters> collection
+* Attribute on type
+
+The difference here is that a custom converter in the `Converters` collection overrides an attribute at the type level. The intention behind this order of precedence is to make run-time changes override design-time choices. There's no way to change the precedence.
+
+### Character escaping
+
+During serialization, `Newtonsoft.Json` is relatively permissive about letting characters through without escaping them. (That is, it doesn't replace them with `\uxxxx` where `xxxx` is the character's code point). Where it does escape them, it does so by emitting a `\` before the character (for example, `"` becomes `\"`). <xref:System.Text.Json> escapes more characters by default to provide defense-in-depth protections against cross-site scripting (XSS) or information-disclosure attacks and does so by using the six-character sequence. `System.Text.Json` escapes all non-ASCII characters by default, so you don't need to do anything if you're using `StringEscapeHandling.EscapeNonAscii` in `Newtonsoft.Json`. `System.Text.Json` also escapes HTML-sensitive characters, by default. For information about how to override the default `System.Text.Json` behavior, see [Customize character encoding](system-text-json-how-to.md#customize-character-encoding).
+
+### Deserialization of Object properties
+
+When `Newtonsoft.Json` deserializes to `Object` properties in POCOs or in dictionaries of type `Dictionary<string, Object>`, it:
+
+* Infers the type of primitive values in the JSON payload (other than `null`) and returns the stored `string`, `long`, `double`, `boolean`, or `DateTime` as a boxed object. *Primitive values* are single JSON values such as a JSON number, string, true, false, or null.
+* Returns the stored complex values as a `JObject` or `JArray`. *Complex values* are collections of JSON key-value pairs within braces (`{}`) or lists of values within brackets (`[]`). The properties and values within the braces or brackets can have additional properties or values.
+* Returns a null reference when the payload has the `null` JSON literal.
+
+<xref:System.Text.Json> stores a boxed `JsonElement` for both primitive and complex values within the `System.Object` property or dictionary value. However, it treats `null` the same as `Newtonsoft.Json` and returns a null reference when the payload has the `null` JSON literal in it.
+
+To implement type inference for `Object` properties, create a converter like the example in [How to write custom converters](system-text-json-converters-how-to.md#deserialize-inferred-types-to-object-properties).
+
+### Maximum depth
+
+`Newtonsoft.Json` doesn't have a maximum depth limit. For <xref:System.Text.Json> there's a default limit  of 64, and it's configurable by setting (xref:System.Text.Json.JsonSerializerOptions.MaxDepth?displayProperty=nameWithType).
+
+### Stack type handling
```

Comment:
This is a timely addition :)
https://stackoverflow.com/questions/59604609/how-can-i-serialize-a-stackt-to-json-using-system-text-json-without-reversing

@layomia, @steveharter - let's make sure folks can workaround this and provide a converter.
https://github.com/dotnet/corefx/issues/41887#issuecomment-571225552

Reply:
Here's a converter for `Stack<T>`:

```C#
public class JsonConverterFactoryForStackOfT : JsonConverterFactory
{
    public override bool CanConvert(Type typeToConvert)
    {
        return typeToConvert.IsGenericType && typeToConvert.GetGenericTypeDefinition() == typeof(Stack<>);
    }

    public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializerOptions options)
    {
        Debug.Assert(typeToConvert.IsGenericType && typeToConvert.GetGenericTypeDefinition() == typeof(Stack<>));

        Type elementType = typeToConvert.GetGenericArguments()[0];

        JsonConverter converter = (JsonConverter)Activator.CreateInstance(
            typeof(JsonConverterForStackOfT<>).MakeGenericType(new Type[] { elementType }),
            BindingFlags.Instance | BindingFlags.Public,
            binder: null,
            args: null,
            culture: null)!;

        return converter;
    }
}

public class JsonConverterForStackOfT<T> : JsonConverter<Stack<T>>
{
    public override Stack<T> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        if (reader.TokenType != JsonTokenType.StartArray || !reader.Read())
        {
            throw new JsonException();
        }

        var elements = new Stack<T>();

        while (reader.TokenType != JsonTokenType.EndArray)
        {
            elements.Push(JsonSerializer.Deserialize<T>(ref reader, options));

            if (!reader.Read())
            {
                throw new JsonException();
            }
        }

        return elements;
    }

    public override void Write(Utf8JsonWriter writer, Stack<T> value, JsonSerializerOptions options)
    {
        writer.WriteStartArray();

        var reversed = new Stack<T>(value);

        foreach (T item in reversed)
        {
            JsonSerializer.Serialize(writer, item, options);
        }

        writer.WriteEndArray();
    }
}
```
That said, `Newtonsoft.Json` and `System.Text.Json` behave identically in this regard and hence this section should be removed from the migration page and added to the converter samples page (under a heading like: “Roundtrip Stack and Stack<T> types by reversing items on serialization”).

```C#
private static void StackBehavior()
{
    Stack<int> stack = JsonSerializer.Deserialize<Stack<int>>("[1, 2, 3]");
    Console.WriteLine(stack.Peek()); // top of stack is 3

    string serialized = JsonSerializer.Serialize(stack);
    Console.WriteLine(serialized); // stack contents are reversed on serialize [3, 2, 1]

    // This matches Newtonsoft.Json behavior
    stack = JsonConvert.DeserializeObject<Stack<int>>("[1, 2, 3]");
    Console.WriteLine(stack.Peek()); // top of stack is 3

    serialized = JsonConvert.SerializeObject(stack);
    // stack contents are reversed on serialize even on Newtonsoft.Json [3, 2, 1]
    Console.WriteLine(serialized);

    // non-generic stack
    var s = new Stack();
    s.Push(1);
    s.Push(2);
    s.Push(3);

    serialized = JsonSerializer.Serialize(s);
    Console.WriteLine(serialized); // stack contents are reversed on serialize, by default [3, 2, 1]

    ImmutableStack<int> y = ImmutableStack.Create(1, 2, 3);
    serialized = JsonSerializer.Serialize(y);
    Console.WriteLine(serialized); // stack contents are reversed on serialize, by default [3, 2, 1]

    // How to use the JsonConverterFactory to get roundtrip behavior.
    var options = new JsonSerializerOptions
    {
        Converters = { new JsonConverterFactoryForStackOfT() },
    };
    stack = JsonSerializer.Deserialize<Stack<int>>("[1, 2, 3]", options);
    Console.WriteLine(stack.Peek()); // top of stack is 3

    serialized = JsonSerializer.Serialize(stack, options);
    Console.WriteLine(serialized); // round-trips exactly [1, 2, 3]
}
```